File: backend/app/routers/expenses.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
from typing import List, Optional

from app.db import get_db_session
from app.models import Expense
from app.schemas import ExpenseCreate, Expense as ExpenseRead, ExpenseUpdate, ExpenseCopyRequest
from app.routers.auth import get_current_user
from app.schemas import User


router = APIRouter()


@router.post("/expenses", response_model=ExpenseRead)
def create_expense(
    payload: ExpenseCreate,
    db: Session = Depends(get_db_session),
    current_user: User = Depends(get_current_user)
):
    """Create a new expense."""
    try:
        # Create the expense with the provided data
        expense = Expense(
            user_id=current_user.id,
            name=payload.name,
            expense_type=payload.expense_type,  # Already mapped in the frontend
            amount=payload.amount,
            category=payload.category,  # Added category field
            start_year=payload.start_year,
            end_year=payload.end_year,
            expected_growth_rate=payload.expected_growth_rate,
            is_tax_deductible=payload.is_tax_deductible,
            notes=payload.notes,
            family_member_id=payload.family_member_id
        )
        db.add(expense)
        db.commit()
        db.refresh(expense)
        return expense
    except Exception as e:
        db.rollback()
        logger.error(f"Error creating expense: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Error creating expense: {str(e)}"
        )


@router.get("/expenses", response_model=List[ExpenseRead])
def list_expenses(
    family_member_id: Optional[int] = None,
    expense_type: Optional[str] = None,
    year: Optional[int] = None,
    db: Session = Depends(get_db_session),
    current_user: User = Depends(get_current_user)
):
    """Get all expenses, with optional filters."""
    query = db.query(Expense).filter(Expense.user_id == current_user.id)
    
    if family_member_id:
        query = query.filter(Expense.family_member_id == family_member_id)
    
    if expense_type:
        query = query.filter(Expense.expense_type == expense_type)
    
    if year:
        # Filter expenses that are active in the given year (start_year <= year and (end_year is None or end_year >= year))
        query = query.filter(Expense.start_year <= year)
        query = query.filter((Expense.end_year.is_(None)) | (Expense.end_year >= year))
    
    return query.all()


@router.get("/expenses/{expense_id}", response_model=ExpenseRead)
def get_expense(
    expense_id: int,
    db: Session = Depends(get_db_session),
    current_user: User = Depends(get_current_user)
):
    """Get a specific expense by ID."""
    expense = db.query(Expense).filter(
        Expense.id == expense_id,
        Expense.user_id == current_user.id
    ).first()
    
    if not expense:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Expense not found"
        )
    
    return expense


@router.put("/expenses/{expense_id}", response_model=ExpenseRead)
def update_expense(
    expense_id: int,
    payload: ExpenseUpdate,
    db: Session = Depends(get_db_session),
    current_user: User = Depends(get_current_user)
):
    """Update an expense."""
    expense = db.query(Expense).filter(
        Expense.id == expense_id,
        Expense.user_id == current_user.id
    ).first()
    
    if not expense:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Expense not found"
        )
    
    # Update fields
    for field, value in payload.dict(exclude_unset=True).items():
        setattr(expense, field, value)
    
    db.commit()
    db.refresh(expense)
    return expense


@router.delete("/expenses/{expense_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_expense(
    expense_id: int,
    db: Session = Depends(get_db_session),
    current_user: User = Depends(get_current_user)
):
    """Delete an expense."""
    expense = db.query(Expense).filter(
        Expense.id == expense_id,
        Expense.user_id == current_user.id
    ).first()
    
    if not expense:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Expense not found"
        )
    
    db.delete(expense)
    db.commit()
    return None


@router.post("/expenses/{expense_id}/copy", response_model=List[ExpenseRead])
def copy_expense_to_years(
    expense_id: int,
    payload: ExpenseCopyRequest,
    db: Session = Depends(get_db_session),
    current_user: User = Depends(get_current_user)
):
    """Copy an expense to multiple years."""
    source_expense = db.query(Expense).filter(
        Expense.id == expense_id,
        Expense.user_id == current_user.id
    ).first()
    
    if not source_expense:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Source expense not found"
        )
    
    new_expenses = []
    for year in payload.target_years:
        # Ensure we're not creating a duplicate for the existing expense's year
        if source_expense.start_year == year:
            continue
            
        # Create a new expense for each target year
        new_expense = Expense(
            user_id=current_user.id,
            name=source_expense.name,
            expense_type=source_expense.expense_type,
            amount=payload.adjust_amount if payload.adjust_amount is not None else source_expense.amount,
            start_year=year,
            end_year=payload.end_year,
            expected_growth_rate=source_expense.expected_growth_rate,
            is_tax_deductible=source_expense.is_tax_deductible,
            notes=source_expense.notes,
            family_member_id=source_expense.family_member_id
        )
        db.add(new_expense)
        new_expenses.append(new_expense)
    
    db.commit()
    
    # Refresh all new expenses
    for expense in new_expenses:
        db.refresh(expense)
    
    return new_expenses 

File: frontend/src/api/expenses.ts
import { api } from './api';

export enum ExpenseTypeEnum {
  HOUSING = "HOUSING",
  TRANSPORTATION = "TRANSPORTATION",
  FOOD = "FOOD",
  UTILITIES = "UTILITIES",
  HEALTHCARE = "HEALTHCARE",
  INSURANCE = "INSURANCE",
  ENTERTAINMENT = "ENTERTAINMENT",
  TRAVEL = "TRAVEL",
  EDUCATION = "EDUCATION",
  SPECIAL = "SPECIAL",
  OTHER = "OTHER"
}

// Backend model
export interface BackendExpense {
  id: number;
  name: string;
  amount: number;
  expense_type: string;
  category?: string;
  family_member_id: number | null;
  start_year: number;
  end_year?: number | null;
  expected_growth_rate?: number;
  is_tax_deductible?: boolean;
  notes?: string | null;
  created_at: string;
  updated_at: string;
}

// Frontend model
export interface Expense {
  id: number;
  name: string;
  amount: number;
  type: string;
  category: string;
  year: number;
  family_member_id: number | null;
  family_member_name?: string | null;
  end_year?: number | null;
  expected_growth_rate?: number;
  is_tax_deductible?: boolean;
  notes?: string | null;
  created_at?: string;
  updated_at?: string;
}

// Transform backend model to frontend model
function transformExpense(backendExpense: BackendExpense): Expense {
  return {
    id: backendExpense.id,
    name: backendExpense.name,
    amount: backendExpense.amount,
    type: backendExpense.expense_type,
    category: backendExpense.category || '',
    year: backendExpense.start_year,
    family_member_id: backendExpense.family_member_id,
    end_year: backendExpense.end_year,
    expected_growth_rate: backendExpense.expected_growth_rate,
    is_tax_deductible: backendExpense.is_tax_deductible,
    notes: backendExpense.notes,
    created_at: backendExpense.created_at,
    updated_at: backendExpense.updated_at
  };
}

export interface ExpenseCreate {
  name: string;
  expense_type: string;
  amount: number;
  category: string;
  family_member_id?: number | null;
  start_year: number;
  end_year?: number | null;
  expected_growth_rate?: number;
  is_tax_deductible?: boolean;
  notes?: string | null;
}

export interface ExpenseUpdate {
  name?: string;
  expense_type?: string;
  amount?: number;
  category?: string;
  family_member_id?: number | null;
  start_year?: number;
  end_year?: number | null;
  expected_growth_rate?: number;
  is_tax_deductible?: boolean;
  notes?: string | null;
}

export interface ExpenseCopy {
  years: number[];
}

export const expenseApi = {
  /**
   * Get all expenses with optional filters
   */
  getAll: async (
    familyMemberId?: number,
    expenseType?: string,
    year?: number
  ): Promise<Expense[]> => {
    let url = '/expenses';
    const params = [];
    
    if (familyMemberId) params.push(`family_member_id=${familyMemberId}`);
    if (expenseType) params.push(`expense_type=${expenseType}`);
    if (year) params.push(`year=${year}`);
    
    if (params.length > 0) {
      url += `?${params.join('&')}`;
    }
    
    const response = await api.get<BackendExpense[]>(url);
    // Transform backend model to frontend model
    return response.map(transformExpense);
  },

  /**
   * Get a single expense by ID
   */
  getById: async (id: number): Promise<Expense> => {
    const response = await api.get<BackendExpense>(`/expenses/${id}`);
    return transformExpense(response);
  },

  /**
   * Create a new expense
   */
  create: async (data: ExpenseCreate): Promise<Expense> => {
    const response = await api.post<BackendExpense>('/expenses', data);
    return transformExpense(response);
  },

  /**
   * Update an existing expense
   */
  update: async (id: number, data: ExpenseUpdate): Promise<Expense> => {
    const response = await api.put<BackendExpense>(`/expenses/${id}`, data);
    return transformExpense(response);
  },

  /**
   * Delete an expense
   */
  delete: async (id: number): Promise<void> => {
    return api.delete(`/expenses/${id}`);
  },

  /**
   * Copy an expense to multiple years
   */
  copyToYears: async (id: number, data: ExpenseCopy): Promise<Expense[]> => {
    const response = await api.post<BackendExpense[]>(`/expenses/${id}/copy`, data);
    return response.map(transformExpense);
  }
}; 